walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,101 @@
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+  // comparator using the Java 8 lambda syntax
+  private final Comparator<Event> ORDER_END =
+      (Event a, Event b) -> Long.compare(a.getWhen().end(), b.getWhen().end());
+
+  public Collection<TimeRange> query(Collection<Event> events,
+                                     MeetingRequest request) {
+    /* Algorithm overview:
+    Remove people from the event list who are in the list but not in the
+    meeting. Sort all of events in descending order by end time. Grab the start
+    time for this current event and traverse through array all events that have
+    an end time greater than this start time. The time slot that can not be used
+    is then the end time of the current iteration and the minimum start time of
+    those other events. Now take the event that after the last event that was
+    discovered still within the constraints and repeat from there. Such an
+    approach will take O(nlogn) time to sort and O(n) time to iterate. The
+    removal will take O(mn) time giving us a complexity of O(mn
+    + nlogn).
+    */
+
+    // Store the times when we cannot meet, essentially take the inverse of this
+    // to determine the times we can meet.
+
+    Collection<String> meetingAttendees = request.getAttendees();
+    ArrayList<TimeRange> fullTimes = new ArrayList<>();
+    Collection<TimeRange> toReturn = new ArrayList<>();
+    // Change to ArrayList so we can sort.
+    ArrayList<Event> eventList = new ArrayList<>(events);
+    // Remove an event if none of the members in the meeting attended it.
+    eventList.removeIf(event
+                       -> !event.getAttendees().stream().anyMatch(
+                           attendee -> meetingAttendees.contains(attendee)));
+    Collections.sort(eventList, ORDER_END.reversed());
+
+    ListIterator<Event> it = eventList.listIterator();
+    while (it.hasNext()) {
+      Event currentEvent = it.next();
+      TimeRange currentTime = currentEvent.getWhen();
+      int currentEndTime = currentTime.end();
+      int currentStartTime = currentTime.start();
+
+      int minStartTime = currentStartTime;
+      while (it.hasNext()) {
+        Event nextEvent = it.next();
+        TimeRange nextEventTime = nextEvent.getWhen();
+        if (nextEventTime.end() >= currentStartTime) {
+          minStartTime = Math.min(currentStartTime, nextEventTime.start());
+        } else {
+          // In this case it is not possible to expand this event any farther,
+          // the iterator already points to the next element we should continue
+          // from.
+          // reset the iterator so the last tried value isn't skipped
+          it.previous();
+          break;
+        }
+      }
+
+      int fullDuration = currentEndTime - minStartTime;
+      // TODO(morleyd): Figure out why the TimeRange class constructor was
+      // defined as private, this seems confusing?
```

Comment:
My guess is: to avoid confusion over the meaning of the constructor parameters. It would be easy to see `TimeRange(10, 30)` in code and mistakenly think it's 20 minutes long. The named factory functions make it harder to make that kind of mistake.

Reply:
That makes sense, I've never seen this done, but your explanation gives a good reason for doing so.